File: LeetCode/Hao/34SearchForARange.java
public class Solution {
    private int findEle(int[] nums, int target, int method) {
        int left, right;
        left = 0;
        right = nums.length - 1;
        
        while (left <= right) {
            if (left == right) {
                break;
            }
            
            int mid = left + (right - left) / 2;
            
            if (target == nums[mid]) {
                if (method == 0){
                    left = mid;
                    break;
                } else if (method == 1) { // find left.
                    right = mid;
                } else { // find right.
                    left = mid + 1;
                }
                
            } else if (target > nums[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        if (method == 0 && nums[left] != target) {
            return -1;
        } else if (method == 1 && nums[left] < target) {
            return left + 1;
        } else if (method == 2 && nums[right] > target) {
            return left - 1;
        }
        
        return left;
    }
    
    public int[] searchRange(int[] nums, int target) {
        int[] position = new int[2];
        position[0] = position[1] = -1;
        
        if (nums == null || nums.length == 0){
            return position;
        }
    
        if (findEle(nums, target, 0) == -1) {
            return position;
        }
        position[0] = findEle(nums, target, 1);
        position[1] = findEle(nums, target, 2);
        return position;
    }
}

File: LeetCode/Liang/34SearchForARange.cpp
/**
 * It's just like upper_bound and lower_bound [ , )
 **/

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target)
    {
        vector<int> result(2, -1);
        int lower = lowerBound(nums, target);
        int upper = upperBound(nums, target);
        if (lower < upper) {
            result.at(0) = lower;
            result.at(1) = upper - 1;
        }
        return result;
    }
    
private:
    // Find the first element index that is no less than target.
    int lowerBound(const vector<int> &nums, int target)
    {
        if (nums.empty()) {
            return -1;
        }
        
        int begin = 0, end = nums.size() - 1;
        while (begin < end) {
            int mid = begin + (end - begin) / 2;
            if (target <= nums.at(mid)) {
                end = mid;
            } else {
                begin = mid + 1;
            }
        }
        
        // Because of case: [1, 1], 2
        if (target > nums.at(begin)) {
            ++begin;
        }
        return begin;
    }
    
    // Find the first element index that is larget than target.
    int upperBound(const vector<int> &nums, int target)
    {
        if (nums.empty()) {
            return -1;
        }
        
        int begin = 0, end = nums.size() - 1;
        while (begin < end) {
            int mid = begin + (end - begin) / 2;
            if (target >= nums.at(mid)) {
                begin = mid + 1;
            } else {
                end = mid;
            }
        }
        
        // Because of case: [1, 1], 2
        if (target == nums.at(begin)) {
            ++begin;
        }
        
        return begin;
    }
};